### DISCUSSION TITLE
Possible to reference current values of a record in a mutation?

### DISCUSSION THREAD

**Discussion by evanRetool** [hasura/graphql-engine#9071] [Question] [answered]
I'm just getting started with Hasura, and I've been loving how easy it is to set up REST endpoints for basic CRUD operations on a Postgres database.

I'm building some update mutations which take a bunch of variables to update all of the fields on a given record type. Here's an example: 

```
mutation MyMutation($id: String!, $account_admin: Boolean = false, $account_id: String, $completed_onboarding: Boolean = false, $email: String, $first_name: String, $last_name: String) {
  update_users_by_pk(pk_columns: {id: $id}, _set: {account_admin: $account_admin, account_id: $account_id, completed_onboarding: $completed_onboarding, email: $email, first_name: $first_name, last_name: $last_name}) {
    account_admin
    account_id
    completed_onboarding
    email
    first_name
    last_name
  }
}
```

This works great when you provide values for each of the variables, but I'm not sure how to make it work in situations where you provide a subset of fields and would like to leave the other fields alone.

For example, let's say I'm working on a user record where `account_admin=false`. I want to change that to make the user an admin, so I pass `id` (pk for `users`) and `account_admin=true` variables. If I do this, the query will null out all of the other fields on the record where values weren't provided and a default isn't specified.

Is it possible to somehow pass the current values of fields on a record as the default value for a mutation variable? I think that would accomplish the desired behavior, but I can't figure out how to do it!

**Comment by adas98012** [comment]
You may like to use Input Types instead of a large list of input parameters for the mutation.
Here is a reference document that you'll find useful:
https://graphql.org/graphql-js/mutations-and-input-types/

**Comment by adas98012** [comment, marked as answer]
@evanRetool 
Hope, you were able to use Input Type to consolidate all the input variables.

Here is some additional info to help you build your mutation :



```
mutation MyMutation($id: String!, $update: users_set_input!) {
  update_users_by_pk(pk_columns: {id: $id}, _set: $update) {
    account_admin
    account_id
    completed_onboarding
    email
    first_name
    last_name
  }
}
```

Your would pass the following variables (to just set the **account_admin** to true and leave other fields untouched):

```
{
  "id": "myid",
  "update": {
    "admin": true
  }
}
```

However, you would see null values in the other fields if you are using **nullable variables** for those.
Because, nullable variables for which a value is not provided default to null (unless another default value is specified in the mutation)

Hope, the info is sufficient to unblock you. Let us know if you need any additional info.

**Reply by evanRetool** [reply]
Hi there, thank you for your advice!

I tried updating my database schema to set all fields as `nullable = false`. However, when I test out a mutation to update a single field on an existing record with non-null values for all fields, I get a "Not-NULL violation" for those fields where my mutation didn't provide a value. For reference, I'm using Postgres 13.3. Is it possible that my database is the problem?

**Reply by adas98012** [reply]
Hi, you don't need to set nullable = false (as that would require all fields to be provided a value). You can look at the example above and pass just a json payload with fields that require an update.

**Reply by evanRetool** [reply]
Oh, I see what you mean! Thank you for the clarification, it looks like this does exactly what I'm looking for.

Thanks again!
